kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/impl/cache/Reflector.java

```diff
@@ -91,35 +94,57 @@ private synchronized void stopWatcher() {
    *
    * @return a future that completes when the list and watch are established
    */
-  public CompletableFuture<Void> listSyncAndWatch() {
+  public CompletableFuture<Void> listSyncAndWatch(boolean reconnect) {
     if (!running) {
       return CompletableFuture.completedFuture(null);
     }
     Set<String> nextKeys = new ConcurrentSkipListSet<>();
-    return processList(nextKeys, null).thenAccept(result -> {
+    CompletableFuture<Void> theFuture = processList(nextKeys, null).thenCompose(result -> {
       store.retainAll(nextKeys);
       final String latestResourceVersion = result.getMetadata().getResourceVersion();
       lastSyncResourceVersion = latestResourceVersion;
       log.debug("Listing items ({}) for {} at v{}", nextKeys.size(), this, latestResourceVersion);
-      CompletableFuture<Watch> started = startWatcher(latestResourceVersion);
-      if (started != null) {
-        // outside of the lock
-        started.whenComplete((w, t) -> {
-          if (w != null) {
-            if (running) {
-              watching = true;
-            } else {
-              stopWatch(w);
-            }
+      return startWatcher(latestResourceVersion);
+    }).thenAccept(w -> {
+      if (w != null) {
+        if (running) {
+          if (log.isDebugEnabled()) {
+            log.debug("Watch started for {}", Reflector.this);
           }
-        });
+          watching = true;
+        } else {
+          stopWatch(w);
+        }
       }
     });
+    if (reconnect) {
+      theFuture.whenComplete((v, t) -> {
+        if (t != null) {
+          log.warn("listSyncAndWatch failed for {}, will retry", Reflector.this, t);
+          reconnect();
+        } else {
+          retryIntervalCalculator.resetReconnectAttempts();
+        }
+      });
+    }
+    return theFuture;
+  }
+
+  private void reconnect() {
+    if (!running) {
+      return;
```

Comment:
If I understand all the invariants correctly the `stopFuture` should be completed in this block (even if it's unlikely to be reached).